### PR TITLE
new(modern_bpf): Add ppc64le support for modern-bpf

### PR DESCRIPTION
**Pull request by mdafsanhossain**
**What type of PR is this?**
/kind feature

**Any specific area of the project related to this PR?**

/area driver-modern-bpf
/area libscap-engine-modern-bpf
/area libscap

**What this PR does / why we need it**:
This PR adds `PPC64LE` specific changes to allow compiling of modern-bpf on PowerPC architecture. This is to add modern-bpf support for Stackrox on ppc64le.

Output of `make ProbeSkeleton` and `make scap` are attached below.

[scap-build.log](https://github.com/falcosecurity/libs/files/13280153/scap-build.log)
[probeskeleton.log](https://github.com/falcosecurity/libs/files/13280154/probeskeleton.log)

Note:  syscall_compat_* headers were generated using `./build/syscalls-bumper -repo-root /root/libs -overwrite`

Does this PR introduce a user-facing change?:

```release-note
new(modern-bpf): support ppc64le architecture.
```

### CODE REVIEW

**Comment by FedeDP** [comment]
Hi! Thanks for this PR! Is there any plan to support ppc64le also for old bpf probe (and eventually kmod too)?
Btw:
> Note: syscall_compat_* headers were generated using ./build/syscalls-bumper -repo-root /root/libs -overwrite

syscalls-bumper does not support ppc64le though; was its support added locally? Can you open a PR over there too?
I noticed that:
* the report.md changes look weird (it seems like `riscv64` arch was dropped?)
* some syscalls were removed from `syscall_compat_*` and that is weird too

**Comment by mdafsanhossain** [comment]
> Hi! Thanks for this PR! Is there any plan to support ppc64le also for old bpf probe (and eventually kmod too)? Btw:
> 
> > Note: syscall_compat_* headers were generated using ./build/syscalls-bumper -repo-root /root/libs -overwrite
> 
> syscalls-bumper does not support ppc64le though; was its support added locally? Can you open a PR over there too? I noticed that:
> 
> * the report.md changes look weird (it seems like `riscv64` arch was dropped?)
> * some syscalls were removed from `syscall_compat_*` and that is weird too

Yes I added the support locally and built it.

```
diff --git a/main.go b/main.go
index ae9eae9..d085ea1 100644
--- a/main.go
+++ b/main.go
@@ -119,10 +119,10 @@ func initOpts() {
 // key should be consistent with the one used by https://github.com/hrw/syscalls-table/tree/master/tables.
 // Value should be the suffix used by libs/driver/syscall_compat_* headers.
 var supportedArchs = map[string]string{
-       "x86_64":  "x86_64",
-       "arm64":   "aarch64",
-       "s390x":   "s390x",
-       "riscv64": "riscv64",
+       "x86_64":    "x86_64",
+       "arm64":     "aarch64",
+       "s390x":     "s390x",
+       "powerpc64": "ppc64le",
 }
 ```
I removed `riscv64` from the list. I will revert the last commit and regenerate it with `riscv64` included. 
Let me know if the above changes are sufficient to generate correctly?

**Comment by FedeDP** [comment]
> I removed riscv64 from the list. I will revert the last commit and regenerate it with riscv64 included.

Thanks!
>  And, are the above changes sufficient to generate correctly?

Yes, they are!
>  Do you know why some of the syscalls are being removed?

Just double checked! Yep because they were cleaned up by the project that syscalls-bumper uses under the hood: https://github.com/hrw/syscalls-table/commit/079ef06025094930efecefbe05d14ac8b26e876a, https://github.com/hrw/syscalls-table/commit/94b4691bfa78a9a15b860db8d6029b5c39591053, https://github.com/hrw/syscalls-table/commit/046c3b55f8c295fad1e2cfadbb96e5530d846006

**Comment by FedeDP** [comment]
Also, you now need a rebase since a PR with automatic bump of syscall-bumper was merged ;)

**Comment by FedeDP** [comment]
Also, once this is fully in place, mind to also update [README](https://github.com/falcosecurity/libs#drivers-officially-supported-architectures) adding the new supported arch for modern bpf probe as _EXPERIMENTAL_?

**Comment by mdafsanhossain** [comment]
> Also, you now need a rebase since a PR with automatic bump of syscall-bumper was merged ;)

Thanks. I will do that, clean it up and after that update the README too.

File: driver/modern_bpf/helpers/interfaces/attached_programs.h
Comment (FedeDP) [inline driver/modern_bpf/helpers/interfaces/attached_programs.h:18]: I think we just need to avoid name collision, by prefixing these ones with eg: `MODERN_BPF_`.
@Andreagit97 thoughts?

**Comment by mdafsanhossain** [comment]
@FedeDP Is it fine that the syscalls are being removed?

**Comment by FedeDP** [comment]
> Is it fine that the syscalls are being removed?

It is ;)
This LGTM except for the comment on `driver/modern_bpf/helpers/interfaces/attached_programs.h`.

Were you able to run [`drivers_test`](https://github.com/falcosecurity/libs/tree/master/test/drivers) tests on a real ppc64le machine?

File: driver/modern_bpf/helpers/interfaces/attached_programs.h
Comment (FedeDP) [reply driver/modern_bpf/helpers/interfaces/attached_programs.h:18]: @mdafsanhossain here !

Comment (mdafsanhossain) [reply driver/modern_bpf/helpers/interfaces/attached_programs.h:18]: @FedeDP Do you mean something like this? `MODERN_BPF_SYSCALL=0`? I am not sure if I am understanding this correctly.

File: driver/SCHEMA_VERSION
Comment (Andreagit97) [inline driver/SCHEMA_VERSION:1]: we bumped this because new PPM_CODES could be sent from the drivers

File: driver/modern_bpf/helpers/interfaces/attached_programs.h
Comment (FedeDP) [reply driver/modern_bpf/helpers/interfaces/attached_programs.h:18]:
```
enum intrumentation_type
{
               MODERN_BPF_SYSCALL = 0,
               MODERN_BPF_TRACEPOINT = 1,
}
```
and then rename their usages whereveer `sampling_logic` is called.

Comment (Andreagit97) [reply driver/modern_bpf/helpers/interfaces/attached_programs.h:18]: yes I agree, probably we can rename them in `INSTR_TYPE_SYSCALL` and `INSTR_TYPE_TRACEPOINT`

Comment (Andreagit97) [reply driver/modern_bpf/helpers/interfaces/attached_programs.h:18]: uhm I didn't see new messages, if we all agree it's also okay to go with 
```c
enum intrumentation_type
{
               MODERN_BPF_SYSCALL = 0,
               MODERN_BPF_TRACEPOINT = 1,
}
```

Comment (mdafsanhossain) [reply driver/modern_bpf/helpers/interfaces/attached_programs.h:18]: Makes sense now. Thanks :)

**Comment by mdafsanhossain** [comment]
> > Is it fine that the syscalls are being removed?
> 
> It is ;) This LGTM except for the comment on `driver/modern_bpf/helpers/interfaces/attached_programs.h`.
> 
> Were you able to run [`drivers_test`](https://github.com/falcosecurity/libs/tree/master/test/drivers) tests on a real ppc64le machine?

@FedeDP I am running into an issue while trying to run the drivers_test
```
-----------------------------------------------------
- Setup phase
-----------------------------------------------------

* Configure modern BPF probe tests!
* Using buffer dim: 8388608
Unable to open the engine: the specified per-CPU ring buffer dimension (0) is not allowed! Please use a power of 2 and a multiple of the actual page_size (65536)!
```

Steps to reproduce:
```
cmake -DUSE_BUNDLED_DEPS=On -DBUILD_LIBSCAP_MODERN_BPF=On -DENABLE_DRIVERS_TESTS=On -DBUILD_LIBSCAP_GVISOR=Off -DCREATE_TEST_TARGETS=On ..
make drivers_test
sudo ./test/drivers/drivers_test -m
```
Btw there are 2 small bugs in these files which prevents drivers_test compilation.

- https://github.com/mdafsanhossain/libs/blob/ec188eb2f97fa927c961595c05ff74275b057b7b/test/drivers/test_suites/syscall_exit_suite/socketcall_x.cpp#L3149 - `int64_t errno_value = -errno; is not declared`
- https://github.com/mdafsanhossain/libs/blob/e2cdf1e084dabc8b42e69bae83ee49a74096ca1c/test/drivers/test_suites/syscall_exit_suite/send_x.cpp#L15 -`No semicolon after int flags=0`

**Comment by Andreagit97** [comment]
```
-----------------------------------------------------
- Setup phase
-----------------------------------------------------

* Configure modern BPF probe tests!
* Using buffer dim: 8388608
Unable to open the engine: the specified per-CPU ring buffer dimension (0) is not allowed! Please use a power of 2 and a multiple of the actual page_size (65536)!
```

This error is weird, the buffer dim is correct in this line
```
* Using buffer dim: 8388608
```

but in the below line is `0`

```
Unable to open the engine: the specified per-CPU ring buffer dimension (0) is not allowed! 
```

It seems like someone on powerpc clears this `buffer_bytes_dim` :thinking: 

> Steps to reproduce:

Just tried on my machine and it works, moreover these tests run in CI so they are working it seems something specific to powerpc

**Comment by mdafsanhossain** [comment]
> ```
> -----------------------------------------------------
> - Setup phase
> -----------------------------------------------------
> 
> * Configure modern BPF probe tests!
> * Using buffer dim: 8388608
> Unable to open the engine: the specified per-CPU ring buffer dimension (0) is not allowed! Please use a power of 2 and a multiple of the actual page_size (65536)!
> ```
> 
> This error is weird, the buffer dim is correct in this line
> 
> ```
> * Using buffer dim: 8388608
> ```
> 
> but in the below line is `0`
> 
> ```
> Unable to open the engine: the specified per-CPU ring buffer dimension (0) is not allowed! 
> ```
> 
> It seems like someone on powerpc clears this `buffer_bytes_dim` 🤔
> 
> > Steps to reproduce:
> 
> Just tried on my machine and it works, moreover these tests run in CI so they are working it seems something specific to powerpc

I wonder what is resetting `buffer_bytes_dim` to 0. I will try to debug it but I don't have much idea on it.

**Comment by Andreagit97** [comment]
> I wonder what is resetting buffer_bytes_dim to 0. I will try to debug it but I don't have much idea on it.

The funny fact is that nobody should touch that value :/ 

```diff
diff --git a/test/drivers/start_tests.cpp b/test/drivers/start_tests.cpp
index 68a98b619..ff6f628f2 100644
--- a/test/drivers/start_tests.cpp
+++ b/test/drivers/start_tests.cpp
@@ -196,6 +196,7 @@ int open_engine(int argc, char** argv)
                        abort_if_already_configured(vtable);
                        vtable = &scap_modern_bpf_engine;
                        modern_bpf_params.buffer_bytes_dim = buffer_bytes_dim;
+                       printf("[BEFORE] buffer bytes dim in the modern: %ld\n", modern_bpf_params.buffer_bytes_dim);
                        oargs.engine_params = &modern_bpf_params;
                        std::cout << "* Configure modern BPF probe tests!" << std::endl;
                }
diff --git a/userspace/libpman/src/configuration.c b/userspace/libpman/src/configuration.c
index 223618bcb..884b4df3e 100644
--- a/userspace/libpman/src/configuration.c
+++ b/userspace/libpman/src/configuration.c
@@ -42,7 +42,7 @@ static int libbpf_print(enum libbpf_print_level level, const char* format, va_li
        }
 
        if(g_state.log_fn == NULL)
-               return vfprintf(stderr, format, args);
+               return 0;
 
        char buf[4096];
        int rc = vsnprintf(buf, sizeof(buf), format, args);
diff --git a/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c b/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c
index 990043354..b08129e0e 100644
--- a/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c
+++ b/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c
@@ -173,6 +173,7 @@ int32_t scap_modern_bpf__init(scap_t* handle, scap_open_args* oargs)
        struct scap_engine_handle engine = handle->m_engine;
        struct scap_modern_bpf_engine_params* params = oargs->engine_params;
 
+       printf("[AFTER] buffer bytes dim in the modern: %ld\n", params->buffer_bytes_dim);
        pman_clear_state();
 
        /* Some checks to test if we can use the modern BPF probe
```

Maybe you could try to add some prints , for example when is set to see if at least is set to the correct value, and when it is received, in the flow no one should touch the value. the `return 0` is useful to disable some logs

**Comment by mdafsanhossain** [comment]
> sudo ./test/drivers/drivers_test -m



> > I wonder what is resetting buffer_bytes_dim to 0. I will try to debug it but I don't have much idea on it.
> 
> The funny fact is that nobody should touch that value :/
> 
> ```diff
> diff --git a/test/drivers/start_tests.cpp b/test/drivers/start_tests.cpp
> index 68a98b619..ff6f628f2 100644
> --- a/test/drivers/start_tests.cpp
> +++ b/test/drivers/start_tests.cpp
> @@ -196,6 +196,7 @@ int open_engine(int argc, char** argv)
>                         abort_if_already_configured(vtable);
>                         vtable = &scap_modern_bpf_engine;
>                         modern_bpf_params.buffer_bytes_dim = buffer_bytes_dim;
> +                       printf("[BEFORE] buffer bytes dim in the modern: %ld\n", modern_bpf_params.buffer_bytes_dim);
>                         oargs.engine_params = &modern_bpf_params;
>                         std::cout << "* Configure modern BPF probe tests!" << std::endl;
>                 }
> diff --git a/userspace/libpman/src/configuration.c b/userspace/libpman/src/configuration.c
> index 223618bcb..884b4df3e 100644
> --- a/userspace/libpman/src/configuration.c
> +++ b/userspace/libpman/src/configuration.c
> @@ -42,7 +42,7 @@ static int libbpf_print(enum libbpf_print_level level, const char* format, va_li
>         }
>  
>         if(g_state.log_fn == NULL)
> -               return vfprintf(stderr, format, args);
> +               return 0;
>  
>         char buf[4096];
>         int rc = vsnprintf(buf, sizeof(buf), format, args);
> diff --git a/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c b/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c
> index 990043354..b08129e0e 100644
> --- a/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c
> +++ b/userspace/libscap/engine/modern_bpf/scap_modern_bpf.c
> @@ -173,6 +173,7 @@ int32_t scap_modern_bpf__init(scap_t* handle, scap_open_args* oargs)
>         struct scap_engine_handle engine = handle->m_engine;
>         struct scap_modern_bpf_engine_params* params = oargs->engine_params;
>  
> +       printf("[AFTER] buffer bytes dim in the modern: %ld\n", params->buffer_bytes_dim);
>         pman_clear_state();
>  
>         /* Some checks to test if we can use the modern BPF probe
> ```
> 
> Maybe you could try to add some prints , for example when is set to see if at least is set to the correct value, and when it is received, in the flow no one should touch the value. the `return 0` is useful to disable some logs

I added the prints. Indeed, it is being set to the correct value but changes to 0 later.

```
-----------------------------------------------------
- Setup phase
-----------------------------------------------------

[BEFORE] buffer bytes dim in the modern: 8388608
* Configure modern BPF probe tests!
* Using buffer dim: 8388608
[AFTER] buffer bytes dim in the modern: 0
Unable to open the engine: the specified per-CPU ring buffer dimension (0) is not allowed! Please use a power of 2 and a multiple of the actual page_size (65536)!
```

**Comment by Andreagit97** [comment]
oh i see just now, in a recent refactor we introduced a bug :face_exhaling: 
```cpp
struct scap_modern_bpf_engine_params modern_bpf_params = {0};
```
should be initialized outside the the switch case otherwise we will point to invalid memory, i will provide the fix! thank you for spotting this!

**Comment by mdafsanhossain** [comment]
> oh i see just now, in a recent refactor we introduced a bug 😮‍💨
> 
> ```c++
> struct scap_modern_bpf_engine_params modern_bpf_params = {0};
> ```
> 
> should be initialized outside the switch case otherwise we will point to invalid memory, i will provide the fix! thank you for spotting this!

I see. Thanks, I will wait for your fix :)

**Comment by Andreagit97** [comment]
This PR improves a little bit the logs and should fix the issue https://github.com/falcosecurity/libs/pull/1477

**Comment by Andreagit97** [comment]
if you rebase on master, it should be fixed!

**Comment by mdafsanhossain** [comment]
@Andreagit97 @FedeDP I am attaching the test result. I see 18 failures.
[drivers-test.log](https://github.com/falcosecurity/libs/files/13305055/drivers-test.log)

**Comment by hbrueckner** [comment]
Let me also have a look at this PR in the next days.

**Comment by Andreagit97** [comment]
@mdafsanhossain thank you for reporting it! Unfortunately without a `PPC64LE` machine is not easy for us to understand what is going on... we could also merge this first PR since the changes are okay and then try to fix tests in a follow-up PR, but let's see what other maintainers think about it!

**Comment by FedeDP** [comment]
/hold for @hbrueckner review :)

> UUnfortunately without a PPC64LE machine is not easy for us to understand what is going on

Agree. Not much we can do to help!

**Comment by mdafsanhossain** [comment]
Thanks guys, that makes sense. I will take a deeper look at the failures in the meantime. We can merge it once @hbrueckner approves.

File: driver/modern_bpf/helpers/extract/extract_from_kernel.h
Comment (FedeDP) [inline driver/modern_bpf/helpers/extract/extract_from_kernel.h:1]: Do we need to also update `bpf_in_ia32_syscall`?

Comment (mdafsanhossain) [reply driver/modern_bpf/helpers/extract/extract_from_kernel.h:1]: I believe yes. Will make the change

**Review by Andreagit97 — APPROVED** [review]
For me it's great, thank you
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Comment by hbrueckner** [comment]
too late.... however, this looks very nice! Thanks a lot!
/approve

**Pull request merged**
